File: backend/app/schemas/favorites.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class FavoriteCreate(BaseModel):
    item_type: str = Field(pattern="^(apod|asteroid|mars_photo|earth_event)$")
    item_payload: dict[str, Any]


class FavoriteResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    item_type: str
    item_payload: dict[str, Any]
    saved_at: datetime

    model_config = {"from_attributes": True}


class FavoriteListResponse(BaseModel):
    favorites: list[FavoriteResponse]
    total: int

File: frontend/stellar-command/src/lib/api/favorites.ts
import { api } from "./client";

export interface Favorite {
  id: string;
  user_id: string;
  item_type: "apod" | "asteroid" | "mars_photo" | "earth_event";
  item_payload: Record<string, unknown>;
  saved_at: string;
}

export interface FavoriteListResponse {
  favorites: Favorite[];
  total: number;
}

export function getFavorites(): Promise<FavoriteListResponse> {
  return api.get<FavoriteListResponse>("/favorites");
}

export function addFavorite(
  item_type: Favorite["item_type"],
  item_payload: Record<string, unknown>,
): Promise<Favorite> {
  return api.post<Favorite>("/favorites", { item_type, item_payload });
}

export function removeFavorite(id: string): Promise<void> {
  return api.delete<void>(`/favorites/${id}`);
}
